### DISCUSSION TITLE
proxy configuration being loaded from trigger in app  i.e with Controller

### DISCUSSION THREAD

**Discussion by cristyonathan** [dotnet/yarp#1644] [Q&A] [answered]
I have question, could YARP load proxy configuration without restart the application?  i have read the documentation about the proxy configuration but either use  LoadFromMemory or LoadFromConfig this way requires yout to restart the application, even though in my case i need to load new proxy config while app is still run i.e use trigger from controller

**Comment by greenEkatherine** [comment]
If you are using configuration files, you don't need to restart YARP after making changes, they are applied automatically. If you are using config provider, you need to set up reload https://microsoft.github.io/reverse-proxy/articles/config-providers.html#reload

**Reply by cristyonathan** [reply]
thanks for the answer, but what really i want in here is user can add or remove proxy configuration by user trigger in example using api controller to maintain the config, it is posible?

**Reply by bmatatuk** [reply]
> If you are using configuration files, you don't need to restart YARP after making changes, they are applied automatically. If you are using config provider, you need to set up reload https://microsoft.github.io/reverse-proxy/articles/config-providers.html#reload

Hi,
How long would it take for it to reload the configuration file changes? FYI: My RP app is running as an Azure App Service (Linux Instance). I upload my config changes using SFTP.

Regards

**Reply by Tratcher** [reply]
It should update immediately. You can check the logs, it will say when it notices and applies the changes.

**Reply by bmatatuk** [reply]
There are no entries with the said changes in the Log Stream on the Azure App Service. It doesn't seem to be applying the changes or detecting the changes whilst the app is running. 

**What I did !!!**
I have added a new route to appsettings.json and uploaded it App Sevice using SFTP and tried for almost 30 minutes at regular interval to see the effect. But nothing. I keeps finding the default or catchall route defined in the setting file. 

It will work if I restart the service, but I can't do that during the normal working hours.

**INFO:**
Azure App Service (Linux Instance)
YARP version : 1.1.1

Thanks

**Reply by Tratcher** [reply]
I suspect this is an issue with IConfiguration change notifications not notifying YARP of the file change. At the start of your app hook into https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.configuration.iconfiguration.getreloadtoken?view=dotnet-plat-ext-7.0#microsoft-extensions-configuration-iconfiguration-getreloadtoken to log if the file change notifications are triggering.

**Reply by bmatatuk** [reply]
The plot thinkens a bit here. When the app is run locally through VS2019, under the console it DOES detect setting change.
![image](https://github.com/microsoft/reverse-proxy/assets/17520551/e28231c6-b2d8-4bf4-98f8-18b908c5e50f)

One other thing i've noticed is, every time I change the appsetting.json, it would trigger 2 entries in the console. Possibly the loading happens twice per change.  See the image attached showing 4 log entries for the 2 changes i've made. 

I main is issue YARP on Azure as Azure App Service. I think when running on Azure as an App Service  (linux) it is failing to detect the config.

**Reply by Tratcher** [reply]
I think those are both issues with file change detection which are managed by the IConfiguration layer. If you reduce your repro to just checking if IConfiguration's reload token fires then the folks over at https://github.com/dotnet/runtime should be able to help more.

**Reply by bmatatuk** [reply]
Hello Tratcher,
I did exactly what you've asked me check and can confirm that the my change detection log entry only appears when running it locally or hosted within IIS. It DOES NOT detect the change when hosted as an Azure App Service (Linux).  I am going to report it to dotnet team as you've mentioned.

But, in the meantime I need some help in how I can manually force a reload the configuration, prefereably using an API call to a controller. I've tried using the solution mentioned in the main chat, but I am not able get it to work. 
I would appreciate if you or someone could give me a helping hand with that.
All I need is to be able to call an API, let say ../Config/ReloadConfiguration and this should force YARP to reload again from the appsettings.json.

Regards

**My startup.cs for your reference below:**
`
   public class Startup
    {
        public IConfiguration Configuration { get; }
        private readonly ILogger _logger ;

        public Startup(IConfiguration configuration)
        {                     
            using var loggerFactory = LoggerFactory.Create(builder =>
            {
                builder.SetMinimumLevel(LogLevel.Information);
                builder.AddConsole();
                builder.AddEventSourceLogger();
            });

            _logger = loggerFactory.CreateLogger<Startup>(); 
            
            Configuration = configuration;
            _logger.LogInformation("*** In Startup Constructor *** ");
            var tokenChange = Configuration.GetReloadToken();
            
            tokenChange.RegisterChangeCallback(LogConfigChange, default);
            
        }
        private void LogConfigChange(object state)
        {
            _logger.LogInformation("*** Config changed detected *** ");
        }
        public void ConfigureServices(IServiceCollection services)
        {

            var proxyBuilder = services.AddReverseProxy();
            proxyBuilder.LoadFromConfig(Configuration.GetSection("ReverseProxy"));
        }

        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }
            
            app.UseRouting();
      
            
            app.UseEndpoints(endpoints =>
            {
                endpoints.MapReverseProxy();
                
                /*
                endpoints.MapGet("/", async context =>
                {
                    await context.Response.WriteAsync("Hello World!");
                });
                */
                
            });
        }
    }

`
The test result:
![image](https://github.com/microsoft/reverse-proxy/assets/17520551/602ddde0-755f-4d9a-a625-3792eb58a90c)

**Reply by samsp-msft** [reply]
@Tratcher - perhaps we should make ConfigurationConfigManager public. This is another case where being able to wrap it with a custom provider would make these kinds of modifications simpler.

@bmatatuk - if you create your own wrapper for IConfiguration that returns your own reload token, then you can fire that notification when you want. As IConfiguration only has a couple of members, creating your own wrapper should not be significant.

**Reply by bmatatuk** [reply]
Thanks.

I don't fully understand how this could be done. I would appreciate if you could share a sample code.

**Comment by MihaZupan** [comment, marked as answer]
Yes, you can have a custom configuration provider that's controlled by an API controller.

You would have to implement a custom `IProxyConfigProvider` and signal the `IChangeToken` when you want the config to reload.

See https://microsoft.github.io/reverse-proxy/articles/config-providers.html - it includes an example `InMemoryConfigProvider`.

**Reply by cristyonathan** [reply]
thanks for the answer, but could you bring me some example for this case?
im still cant understand how to implement this [example](https://microsoft.github.io/reverse-proxy/articles/config-providers.html) in my controller

**Reply by Tieantono** [reply]
> Yes, you can have a custom configuration provider that's controlled by an API controller.
> 
> You would have to implement a custom `IProxyConfigProvider` and signal the `IChangeToken` when you want the config to reload.
> 
> See https://microsoft.github.io/reverse-proxy/articles/config-providers.html - it includes an example `InMemoryConfigProvider`.

@MihaZupan Does this mean we should have registered both of `IProxyConfigProvider` and our custom one, let's say `CustomInMemoryConfigProvider`, in order for our client codes can invoke the `Update` method for signaling the `IChangeToken`?

Below are my working service registration codes, but I wonder is this the correct approach for this case or not:
```csharp
// If I don't put this line, it would trigger an exception:
// Unable to resolve service for type 'Yarp.ReverseProxy.Configuration.IProxyConfigProvider' while attempting to activate 'Yarp.ReverseProxy.Management.ProxyConfigManager'.
builder.Services.AddSingleton<IProxyConfigProvider>(new CustomInMemoryConfigProvider(routes, clusters));

// Register a custom config provider based on this example: https://microsoft.github.io/reverse-proxy/articles/config-providers.html#example.
// This is needed so our client codes can invoke the implemented Update() method to signal the IChangeToken.
builder.Services.AddSingleton(new CustomInMemoryConfigProvider(routes, clusters));
```

But using the codes above, the new routes and clusters settings doesn't seems reloaded, since my HTTP request using the newly added path still returns `404 Not Found`. Below is my `AddResponseTransform()` codes:

```csharp
proxyBuilder.LoadFromMemory(Routes, Clusters)
            .AddTransforms(builderContext =>
            {
                builderContext.AddResponseTransform(async context =>
                {
                    var proxyFeature = context.HttpContext.GetReverseProxyFeature();
                    var routeId = string.Empty;
                    Console.WriteLine("proxied destination: " + proxyFeature!.ProxiedDestination!.DestinationId);
                    if (proxyFeature != null && string.IsNullOrEmpty(proxyFeature!.Route!.Config!.Transforms!.First()["PathSet"]))
                    {
                        context.SuppressResponseBody = true;
                        context.HttpContext.Response.Headers.Clear();
                        // Hardcoded JSON content-type
                        context.HttpContext.Response.Headers.ContentType = "application/json; charset=utf-8";
                        var clusterId = proxyFeature.Cluster.Config.ClusterId;
                        var defaultResponse = await apusService.GetApiDefaultResponse(clusterId);
                        await context.HttpContext.Response.Body.WriteAsync(Encoding.UTF8.GetBytes(defaultResponse));
                    }
                });
            });
```

**Reply by MihaZupan** [reply]
If you want your `CustomInMemoryConfigProvider` to be the only config provider and not merge it with the configuration file, only register *it* in DI.

When registering your implementation, you have to tell DI what interface you are implementing
```diff
-builder.Services.AddSingleton(new CustomInMemoryConfigProvider(routes, clusters));
+builder.Services.AddSingleton<IProxyConfigProvider>(new CustomInMemoryConfigProvider(routes, clusters));
```
Without that, YARP won't pick it up as a provider, and you'll see the "Unable to resolve service" exception you mentioned.

**Reply by Tieantono** [reply]
<details>
  <summary>Answers</summary>

> If you want your `CustomInMemoryConfigProvider` to be the only config provider and not merge it with the configuration file, only register _it_ in DI.
> 
> When registering your implementation, you have to tell DI what interface you are implementing
> 
> ```diff
> -builder.Services.AddSingleton(new CustomInMemoryConfigProvider(routes, clusters));
> +builder.Services.AddSingleton<IProxyConfigProvider>(new CustomInMemoryConfigProvider(routes, clusters));
> ```
> 
> Without that, YARP won't pick it up as a provider, and you'll see the "Unable to resolve service" exception you mentioned.
</details>


Hi @MihaZupan , thank you for your answer, I understand that we could just register it like what you've answered, but we were unable to invoke the implemented `Update` method in our client codes because we initially thought that we must register and inject `CustomInMemoryConfigProvider` in another service class. After some thoughts, we have implemented these following codes in our client codes, we name it as `ReloadConfigService`:
```csharp
private readonly InMemoryConfigProvider _configProvider;

public ReloadConfigService(IProxyConfigProvider proxyConfigProvider)
{
    _configProvider = (InMemoryConfigProvider)proxyConfigProvider;
}

public void ChangeConfig()
{
    // MapProxyConfigs() is our custom method for obtaining the list of routes and clusters.
    var (Routes, Clusters) = MapProxyConfigs(listAPI);

    _configProvider.Update(Routes, Clusters);
}
```
The above codes works, but I'm still bugged by the forced typecasting in the constructor, or is there any better way to do use the `Update` method?

**Reply by Tratcher** [reply]
How about this?
```
        builder.Services.AddSingleton(new InMemoryConfigProvider(routes, clusters));
        builder.Services.AddSingleton<IProxyConfigProvider>(s => s.GetRequiredService<InMemoryConfigProvider>());
...
        context.RequestServices.GetRequiredService<InMemoryConfigProvider>().Update(GetRoutes(), GetClusters());
```

**Reply by Tieantono** [reply]
> How about this?
> 
> ```
>         builder.Services.AddSingleton(new InMemoryConfigProvider(routes, clusters));
>         builder.Services.AddSingleton<IProxyConfigProvider>(s => s.GetRequiredService<InMemoryConfigProvider>());
> ...
>         context.RequestServices.GetRequiredService<InMemoryConfigProvider>().Update(GetRoutes(), GetClusters());
> ```

Thanks @Tratcher , this one works well.
